SingularityService/src/main/java/com/hubspot/singularity/helpers/TaskLagGuardrail.java

```diff
@@ -0,0 +1,57 @@
+package com.hubspot.singularity.helpers;
+
+import com.google.inject.Inject;
+import com.hubspot.singularity.SingularityPendingTaskId;
+import com.hubspot.singularity.config.SingularityConfiguration;
+import com.hubspot.singularity.data.TaskManager;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+
+public class TaskLagGuardrail {
+  private final TaskManager taskManager;
+  private final SingularityConfiguration configuration;
+  private ConcurrentMap<String, Integer> lateTasksByRequestId;
+  private long lastUpdate;
+
+  @Inject
+  public TaskLagGuardrail(
+    SingularityConfiguration configuration,
+    TaskManager taskManager
+  ) {
+    this.configuration = configuration;
+    this.taskManager = taskManager;
+    this.lateTasksByRequestId = new ConcurrentHashMap<>();
+    this.lastUpdate = 0;
+  }
+
+  public boolean isLagged(String requestId) {
+    updateLateTasksByRequestId();
+    return lateTasksByRequestId.containsKey(requestId);
```

Comment:
I'm a bit worried about this being called in a super tight loop. I know it's mostly going to no-op on the if statement. But, does it make sense to be updating this on a thread that is trying to grab a lock vs just doing it in the background? Then we can be sure every check is just a map.containsKey and never has to worry about heavier calculations on an active path

Reply:
My original thinking was that a scheduled thread would be overkill for this, but you're right that this is going to be a hot method and that I should add a background thread specifically for this.